### DISCUSSION TITLE
Clarifying resource priority

### DISCUSSION THREAD

**Discussion by stuvet** [ropensci/targets#562] [Help]
Hi,

Having learned by playing, I ended up with a messy codebase which had multiple levels of resource definitions. I've cleaned up my code but I was left wondering if the layered resource definitions had led to some of the unexpected behaviour I'd noticed, particularly with apparently premature ending of walltimes even on non-preemptible machines.

Example of 'messy' resource definitions:

```
library(targets)

slurm <- future::tweak(
                 batchtools_slurm, 
                 template = 'batchtools.slurm.tmpl', 
                 workers = 30L,
                 resources = list(
                   walltime = 3600,
                   ncpus = 4,
                   ntasks = 1,
                   memory = 1750,
                   partition = 'default'
                )
             )

future::plan(slurm)

tar_option_set(
         resources = tar_resources(
                aws = tar_resources_aws(bucket = "my_bucket"),
                future = tar_resources_future(
                   plan = future::plan(slurm),
                   resources = list(
                     walltime = 36000,
                     partition = 'pm2',
                     memory = 14400,
                     workers = 10L,
                     ncpus = 1,
                     ntasks = 1
                 )
            )
       )
)



list(
   tar_target(
        my_target,
        Sys.sleep(60),
        deployment = 'worker',
        resources =   tar_resources(
                aws = tar_resources_aws(bucket = "my_bucket"),
                future = tar_resources_future(
                   plan = future::plan(slurm),
                   resources = list(
                     walltime = 72000,
                     partition = 'ph32',
                     memory = 1750,
                     workers = 30L,
                     ncpus = 4,
                     ntasks = 1
                 )
             )
       )
  ) 
)

```

As part of the scourge I removed the `resources` & `workers` keys in the future::tweak statement & am testing at the moment, but could you please clarify (& perhaps in the manual) how targets prioritises layered resource definitions like these? Apologies if I've missed it.

Thanks.

**Comment by wlandau** [comment]
`targets` currently ignores the resources you supply to `future::plan()` and/or `future::tweak()`. (Maybe it should not do that.) The resources supplied to `tar_target()` completely supersede those supplied to `tar_option_set()`. So in your case, `my_target` should use a walltime of 72000, and any target with no resources specified at all will use a walltime of 36000. Also, in `targets`, resource specification is all or nothing. In other words, if you specify some resources in `tar_target()` but omit `walltime`, then the walltime of that target will default to whatever is in the `batchtools` template file, not `tar_option_get("resources")`.

**Reply by wlandau** [reply]
Just added some clarifying comments to the Rd help files.

**Reply by stuvet** [reply]
Perfect. Thanks. That's pretty much what I assumed (though I was on the fence about fully-replaced vs partially-updated specifications). Turns out that doesn't explain the occasional odd walltime, though I've had to make some edits to my batchtools' namespace even to come close to stability on my slurm infrastructure (e.g. [this](https://github.com/mllg/batchtools/pull/274#issue-662253912)) so I can't yet expect perfection.

Edit: odd walltimes could be slurm-level preemption rather than GCP-level preemption.

**Reply by wlandau** [reply]
Glad that helped. 

Related: unfortunately, due to conflicts with the previous deprecated resource system (#504), `targets` cannot safely accept resources defined in `future::plan()` or `future::tweak()`. Maybe we can add that when unstructured resources are totally defunct later on.

**Reply by stuvet** [reply]
Worth knowing for others, but not an issue for me even with my workarounds. Thanks for clarification though.

**Comment by wresch** [comment]
I am a bit confused. Starting out with a very similar setup:

#### versions
```
 other attached packages:                                           
 [1] targets_0.7.0            future.batchtools_0.10.0 future_1.22.1
 ```

#### `_targets.R`
```
library(targets)
library(future.batchtools)

slurm <- future::tweak(batchtools_slurm, template = 'slurm')
future::plan(slurm)

list(
   tar_target(
        my_target,
        Sys.sleep(60),
        deployment = 'worker',
        resources =   tar_resources(
                future = tar_resources_future(
                   plan = future::plan(slurm),
                   resources = list(
                     walltime = 20,
                     partition = 'quick',
                     memory = 1750,
                     ncpus = 4
                 )
             )
       )
  )
)
```

#### `batchtools.slurm.tmpl`
```
#!/bin/bash
## Job Resource Interface Definition
##
## ncpus [integer(1)]:        Number of required cpus per task,
##                            Set larger than 1 if you want to further parallelize
##                            with multicore/parallel within each task.
## walltime [integer(1)]:     Walltime for this job, in minutes.
## memory   [integer(1)]:     Memory in megabytes.
##                            Must be at least 100 (when I tried lower values my
##                            jobs did not start at all).
##
## optional: partition, gres
##
## Default resources can be set in your .batchtools.conf.R by defining the variable
## 'default.resources' as a named list.
<%
# relative paths are not handled well by Slurm
log.file = fs::path_expand(log.file)
-%>
#SBATCH --job-name=<%= job.name %>
#SBATCH --output=<%= log.file %>
#SBATCH --error=<%= log.file %>
#SBATCH --time=<%= resources$walltime %>
#SBATCH --cpus-per-task=<%= resources$ncpus %>
#SBATCH --mem=<%= resources$memory %>
<%= if (!is.null(resources$partition)) sprintf("#SBATCH --partition=%s", resources$partition) -%>
<%= if (!is.null(resources$gres)) sprintf("#SBATCH --gres=%s", resources$gres) -%>
<%= if (array.jobs) sprintf("#SBATCH --array=1-%i", nrow(jobs)) else "" %>

## Initialize work environment like
## source /etc/profile
## module add ...
module load R/4.1

## Run R:
## we merge R output with stdout from SLURM, which gets then logged via --output option
Rscript -e 'batchtools::doJobCollection("<%= uri %>")'
```

Running `tar_make_future(workers=1` however fails with
```
Error : Fatal error occurred: 101. Command 'sbatch' produced exit code 1. Output: 'sbatch: error: Invalid --time specification'
```

And looking at the processed script in the .future directory, I see that apparently the template got handed an empty `resources` list. Am I missing something?

**Reply by wlandau** [reply]
Your usage looks correct. (Except maybe you need `template = "batchtools.slurm.tmpl"` instead of `template = "slurm"`?) I am not sure what the problem could be, but as I say at https://github.com/ropensci/targets/discussions/432#discussioncomment-652360, I recommend peeling back layers and seeing if you can reproduce the problem without `targets`, e.g. just `future` and `future.batchtools`. Using the same package versions as you (and also using https://github.com/HenrikBengtsson/future/commit/344a97f1cc85283095a43fe01cfabf3fb33fdf70 and https://github.com/HenrikBengtsson/future.batchtools/commit/ea3b86cbc6b30f62f78cfd9ef7c5a64f5d42e683) I encountered a similar issue using the following template file on an SGE cluster. (I do not have access to Slurm.) `sge_batchtools.tmpl`:

```bash
#!/bin/bash
#$ -cwd
#$ -j y
#$ -o <%= log.file %>
#$ -V
#$ -N <%= job.name %>
#$ -pe smp <%= resources$slots %>
module load R/4.0.3
Rscript -e 'batchtools::doJobCollection("<%= uri %>")'
exit 0
```

R console:

```r
library(future)
library(future.batchtools)
plan(batchtools_sge, template = "sge_batchtools.tmpl")
future(Sys.sleep(5), resources = list(slots = 2))
#> Error: Fatal error occurred: 101. Command 'qsub' produced exit code 2. Output: 'Unable to
#> read script file because of error: ERROR! -pe option must have range as 2nd argument'
```

But the job does submit correctly with the correct number of slots if I supply `resources` directly to `plan()`.

```r
plan(batchtools_sge, template = "sge_batchtools.tmpl", resources = list(slots = 2))
future(Sys.sleep(5)) # Launches an SGE job with 2 slots.
```

@wresch, would you confirm if you see the same issue on Slurm and then submit an issue to https://github.com/HenrikBengtsson/future.batchtools/issues if you do? I seem to recall this was working the last time I submitted `targets` to CRAN.

cc @HenrikBengtsson

**Reply by wresch** [reply]
"slurm" works as a template name - i verify that it's using my template by looking at the script generated in the `.future` directory. The search order for templates is described at https://mllg.github.io/batchtools/reference/makeClusterFunctionsSlurm.

I can confirm that providing a straight list as resources to the plan 
```
slurm <- future::tweak(batchtools_slurm, template='slurm', resources= list(
                     walltime = 10,
                     partition = 'quick',
                     memory = 1600,
                     ncpus = 4
                 )
)
future::plan(slurm)
```
works for the simple targets test - the job gets submitted and the script in `.future` has the template filled out correctly with the resources provided to plan, *not* `tar_target`, which appears to be ignored.

I can also confirm that future.batchtools by itself works as expected in this scenario:
```
> future::plan(batchtools_slurm, template='slurm', resources=list(walltime=10, partition='quick', memory=1400, ncpus=2))
> x %<-% { Sys.sleep(5); 3.14 }
```

doesn't that suggest that this is an issue with targets not behaving as documented?

**Reply by wlandau** [reply]
Would you try the following instead? If you supply the resources to `tar_target()`, that is equivalent to supplying `resources` to `future::future()` instead of `future::plan()`.

```r
library(future)
library(future.batchtools)
plan(batchtools_slurm, template = "slurm")
future(Sys.sleep(5), resources = list(walltime = 10, partition = "quick", memory = 1400, ncpus = 2))
```

**Reply by wlandau** [reply]
Just posted to `future.batchtools`: https://github.com/HenrikBengtsson/future.batchtools/issues/75

**Reply by wlandau** [reply]
You can temporarily solve the issue by downgrading to `future` version 1.21.0. Worked on my end.

```r
remotes::install_version("future", version = "1.21.0")
```

**Reply by wresch** [reply]
Indeed - in that case the resources are not passed along to the template script. I wasn't aware that it's supposed to be possible to pass resources to future directly. Is that documented somewhere? I'm fairly new to `future`

**Reply by wresch** [reply]
Sorry I missed your posts above before replying. Thank you very much - that clarifies things

**Reply by wresch** [reply]
The downgrade to future 1.21.0  worked and now resources specified in `tar_target` work as expected.  Thank you very much!

**Reply by wlandau** [reply]
Please have a look at Henrik's response to https://github.com/HenrikBengtsson/future.batchtools/issues/75. In subsequent versions of `future`, it looks like temporary `future` plans are the way to accomplish what you are trying to do. I updated https://books.ropensci.org/targets/hpc.html#future-configuration to explain how.
